### PR TITLE
Change profile codes to rely on the Thunderstore API

### PR DESCRIPTION
**Pull request by Oksamies**
I can in no way be held responsible for the quality of this stuff. Forgive me emperor I've become a heretic.

That said, I just kinda slapped together what I think could work and everything seems to work on my dev env. Do let me know if there is something I've missed.

### CODE REVIEW

**Comment by Oksamies** [comment]
> This change doesn't actually match what was specced

Alright, now the changes should reflect the specs pretty good. I didn't separate the changes into a new file since the there are not that much changes and on top of that the wrappers were not needed, cause Axios gives the Promise straight away 👍

File: src/r2mm/profiles/ProfilesClient.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import Axios, { AxiosResponse } from 'axios';
+import R2Error from 'src/model/errors/R2Error';
+
+export default class ProfilesClient {
```
Comment (MythicManiac) [inline src/r2mm/profiles/ProfilesClient.ts:4]: There's no need to use a class here since the file itself is already handling the encapsulation. Classes can be instantiated which doesn't really make sense here.

Also generally speaking, default exports are worse than named exports, so use them instead unless default exports are explicitly required (like with NextJS)

File: src/r2mm/profiles/ProfilesClient.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                endpoint,
+                payload,
+                {
+                    headers: {'Content-Type': 'application/octet-stream'}
+                }
+            )
+        } catch (e: any) {
+            if (Axios.isAxiosError(e)) {
+                if (e.response!.status == 429) {
+                    return new R2Error('Too many exports in a short period of time', ':PeepoShake:', 'Slow down there champ');
```
Comment (MythicManiac) [inline src/r2mm/profiles/ProfilesClient.ts:21]: `:PeepoShake:`?

Comment (MythicManiac) [inline src/r2mm/profiles/ProfilesClient.ts:21]: We should probably just throw these rather than returning them, and add a try/catch to the call site

File: src/r2mm/profiles/ProfilesClient.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                } else {
+                    return new R2Error('Failed to export profile', e.message, `${e.response!.status}`);
+                }
+            } else {
+                return new R2Error('Unexpected error', e.message, null);
```
Comment (MythicManiac) [inline src/r2mm/profiles/ProfilesClient.ts:26]: And since we're throwing them, you can just throw the original error here if we don't do any special handling for it and let the call site deal with it.

File: src/r2mm/profiles/ProfilesClient.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            }
+        }
+    }
+
+    public static async retrieveProfile(endpoint: string): Promise<AxiosResponse> {
+        return await Axios.get(endpoint)
```
Comment (MythicManiac) [inline src/r2mm/profiles/ProfilesClient.ts:32]: Semicolon missing, but also you can remove `async` from the function and `await` from this line and it'll work exactly as it does now (since Axios already returns a `Promise<AxiosResponse>` and you can just pass it forward). Semicolon also missing.

File: src/r2mm/profiles/ProfilesClient.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import Axios, { AxiosResponse } from 'axios';
+import R2Error from 'src/model/errors/R2Error';
+
+export default class ProfilesClient {
+    public static TSCreateAPI = 'https://thunderstore.io/api/experimental/legacyprofile/create/'
+    public static TSRetrieveAPI = (profileImportCode: string) => {return `https://thunderstore.io/api/experimental/legacyprofile/get/${profileImportCode}/`}
+    public static HerokuRetrieveAPI = (profileImportCode: string) => {return `https://r2modman-hastebin.herokuapp.com/raw/${profileImportCode}`}
+
+    public static async sendProfile(endpoint: string, payload: string): Promise<AxiosResponse | R2Error> {
```
Comment (MythicManiac) [inline src/r2mm/profiles/ProfilesClient.ts:9]: Make sure to remove `| R2Error` from this type once you change the error handling to use throws instead of returns.

File: src/r2mm/profiles/ProfilesClient.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import Axios, { AxiosResponse } from 'axios';
+import R2Error from 'src/model/errors/R2Error';
+
+export default class ProfilesClient {
+    public static TSCreateAPI = 'https://thunderstore.io/api/experimental/legacyprofile/create/'
+    public static TSRetrieveAPI = (profileImportCode: string) => {return `https://thunderstore.io/api/experimental/legacyprofile/get/${profileImportCode}/`}
+    public static HerokuRetrieveAPI = (profileImportCode: string) => {return `https://r2modman-hastebin.herokuapp.com/raw/${profileImportCode}`}
```
Comment (MythicManiac) [inline src/r2mm/profiles/ProfilesClient.ts:7]: You don't need to use return here at all if you remove the `{}`. Semicolons missing

File: src/r2mm/profiles/ProfilesClient.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    public static async sendProfile(endpoint: string, payload: string): Promise<AxiosResponse | R2Error> {
+        try {
+            return await Axios.post(
+                endpoint,
+                payload,
+                {
+                    headers: {'Content-Type': 'application/octet-stream'}
+                }
+            )
```
Comment (MythicManiac) [inline src/r2mm/profiles/ProfilesClient.ts:17]: semicolon missing

File: src/r2mm/profiles/ProfilesClient.ts (hunk 6 above)
Comment (MythicManiac) [inline src/r2mm/profiles/ProfilesClient.ts:7]: These could be consts when not in a class

File: src/r2mm/mods/ProfileModList.ts [hunk 8]
```diff
@@ -248,13 +249,12 @@ export default class ProfileModList {
...
         } else {
             await exportBuilder.createZip(exportPath);
-            const profileBuffer = '#r2modman\n' + (await fs.base64FromZip(exportPath));
-            Axios.post('https://r2modman-hastebin.herokuapp.com/documents', profileBuffer)
-                .then(resp => callback(resp.data.key, null))
-                .catch(e => {
-                    const err: Error = e;
-                    callback('', new R2Error('Failed to export profile', err.message, null));
-                });
+            const storage_response_or_error = await ProfilesClient.sendProfile(ProfilesClient.TSCreateAPI, '#r2modman\n' + (await fs.base64FromZip(exportPath)))
```
Comment (MythicManiac) [inline src/r2mm/mods/ProfileModList.ts:252]: We're not writing python so `camelCase` is preferred.

**Review by MythicManiac — CHANGES_REQUESTED** [review]
Better but not quite there.

To summarize:
- Move as much logic away from the vue components as possible. It's likely at least a part of the error handling will have to remain in the components however.
- Add missing semicolons
- Use proper casing with variable names
- Some other refactoring suggestions

File: src/pages/Profiles.vue [hunk 9]
```diff
@@ -478,22 +479,30 @@ export default class Profiles extends Vue {
...
-                if (resp.startsWith("#r2modman")) {
-                    const buf = Buffer.from(resp.substring(9).trim(), 'base64');
-                    await FileUtils.ensureDirectory(path.join(PathResolver.ROOT, '_import_cache'));
-                    await fs.writeFile(path.join(PathResolver.ROOT, '_import_cache', 'import.r2z'), buf);
-                    await this.importProfileHandler([path.join(PathResolver.ROOT, '_import_cache', 'import.r2z')]);
-                } else {
-                        throw new Error('Code invalid, no profile is associated with this code');
-                }
-            }).catch(e => {
+    async parseProfileResponse(response_data: string) {
```
[the remark this answers is not in the document]
Comment (Oksamies) [reply src/pages/Profiles.vue:482]: With a quick glance, seems like doing this would mean splitting right about the whole file down into separate modules. I'll leave it out of this PR, but I'll drag it through the post-mortem and wrap-up so we'll have task for refactoring the mod manager code into the modules; so that it can be used more easily in the future.

File: src/pages/Profiles.vue [hunk 10]
```diff
@@ -478,22 +479,30 @@ export default class Profiles extends Vue {
...
+        } catch (e: any) {
+            try {
+                const resp = await ProfilesClient.retrieveProfile(ProfilesClient.HerokuRetrieveAPI(this.profileImportCode))
+                this.parseProfileResponse(resp.data)
+            } catch (e: any) {
                 const err = new R2Error('Failed to find profile', e.message, null);
                 this.showError(err);
-            })
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Oksamies) [reply src/pages/Profiles.vue:505]: I kind of cant move anything from `importProfileUsingCode` to the `ProfileClient.ts` module, without dragging the whole `Profiles.vue` file with it. Or might be that I don't know how to. Anyways going bit out of scope, from getting the heroku fix done.

I suggest we move forward with this (if there are no big issues with this) and move the "refactoring" to the next milestone.

[the remark this answers is not in the document]
Comment (MythicManiac) [reply src/pages/Profiles.vue]: As mentioned on Discord, as long as the API endpoint fallback logic is in the new module this is fine, so the bare minimum is to move that portion of the logic in the API module.

On another note, don't try catch but simply check if the first returns a 404 status code and use that as a signal to test the fallback. Falling back to the old API if the new API is returning some other error than 404 is wrong behavior and could lead to some edge case bugs (not really in this case but at l east invalid error messages). In fact you probably shouldn't even use try/catch in the API module at all, just check for status codes and return the data as is.

**Comment by anttimaki** [comment]
Well I don't have too much to add to what Mythic already mentioned. Mildly annoyed by the code style things (missing semicolons, mixing single and double quotes). I guess those should be handled by a linter but in the meanwhile wouldn't hurt to be meticulous about them.

File: src/r2mm/profiles/ProfilesClient.ts (hunk 1 above)
Comment (ebkr) [reply src/r2mm/profiles/ProfilesClient.ts]: The general pattern so far has been to use default exports as most things are classes. Named exports tend to be type declarations.

Reasoning is that it's easier to remember a class name than it is to remember a method name which ends up being more productive.

File: src/r2mm/profiles/ProfilesClient.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        payload,
+        {
+            headers: {'Content-Type': 'application/octet-stream'}
+        }
+    );
+}
+
+export async function retrieveProfile(profileImportCode: string): Promise<AxiosResponse> {
+    const instance = Axios.create({
+        validateStatus: function(status) {
```
Comment (ebkr) [inline src/r2mm/profiles/ProfilesClient.ts:19]: Is this needed?

**Review by ebkr — CHANGES_REQUESTED** [review]

**Comment by ebkr** [comment]
@anttimaki The project used to have ESLint however it was incredibly slow.
It would freeze up VSC for ages, so it was disabled.

File: src/r2mm/profiles/ProfilesClient.ts (hunk 11 above)
Comment (Oksamies) [reply src/r2mm/profiles/ProfilesClient.ts:19]: Yes if we don't want to use try catch (mythics suggestion in comments above). Without this Axios will throw an AxiosError and the latter code wont be run. Now that I think about it, if Axios gets something else than a 404 like 500, it probably wont work. I'll change it back to the try catch since it's more reliable.

Comment (MythicManiac) [reply src/r2mm/profiles/ProfilesClient.ts:19]: Try catch is fine to use but the point was not everything has to be wrapped in it since you can simply let the promise fail if something unexpected happens.

File: src/pages/Profiles.vue (hunk 9 above)
[the remark this answers is not in the document]
Comment (MythicManiac) [reply src/pages/Profiles.vue]: Opened #883 regarding this

**Review by ebkr — APPROVED** [review]
Minor comments but approved to not block once fixed

File: src/model/errors/R2Error.ts [hunk 12]
```diff
@@ -5,11 +5,24 @@ export default class R2Error extends Error {
     public solution: string;
     public errorReferenceString: string | undefined;
 
-    public constructor(name: string, message: string, solution: string | null) {
+    public constructor(name: string, message: string, solution: string | null = null) {
```
[the remark this answers is not in the document]
Comment (MythicManiac) [reply src/model/errors/R2Error.ts:8]: This was set to null as that's what it's typed as and I don't want to risk breaking something by changing the signature

File: src/model/errors/R2Error.ts [hunk 13]
```diff
@@ -5,11 +5,24 @@ export default class R2Error extends Error {
...
         super(message);
         this.name = name;
         this.message = message;
         this.solution = solution || '';
         Object.setPrototypeOf(this, R2Error.prototype);
     }
+
+    public static FromThrownValue(
```
[the remark this answers is not in the document]
Comment (MythicManiac) [reply src/model/errors/R2Error.ts:16]: It was intentionally uppercased to differentiate between static class-scoped methods vs. object instance scoped methods (a fairly common naming convention), but I can be change it if you prefer not to make that distinction visible in naming.

File: src/model/errors/R2Error.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (MythicManiac) [reply src/model/errors/R2Error.ts:8]: Marking as resolved as discussed on Discord

**Review by MythicManiac — APPROVED** [review]
